src/memory.jl

```diff
@@ -38,6 +38,40 @@ function free(p::OwnedPtr)
     @apicall(:cuMemFree, (Ptr{Void},), p.ptr)
 end
 
+const _refcount_dictionary = Dict{Ptr{Void}, Int}() # set to make things kinda pure?!
+
+function refcount(ptr::T) where T
+    get(_refcount_dictionary, Base.unsafe_convert(Ptr{Void}, ptr), 0)
+end
+"""
+Registers a new julia reference for `ptr`
+"""
+function create_reference!(ptr::T) where T
+    _ptr = Base.unsafe_convert(Ptr{Void}, ptr)
+    refcount = get!(_refcount_dictionary, _ptr, 0)
+    _refcount_dictionary[_ptr] = refcount + 1
+    return
+end
+"""
+Signals that there is one less reference to `ptr`.
+If reference count will get decreased to 0, `free_func(ptr, args...)` will get called
+"""
+function free_reference!(ptr::T) where T
+    _ptr = Base.unsafe_convert(Ptr{Void}, ptr)
+    if !haskey(_refcount_dictionary, _ptr)
+        error("Freeing unknown reference: $ptr")
+    end
+    refcount = _refcount_dictionary[_ptr]
+    if refcount == 0
+        error("Double free of reference: $ptr")
```

Comment:
maybe make it an `@assert` then, given that it's not a true error but a corrupt internal state